x-pack/plugins/security/server/routes/authentication.ts

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { schema } from '@kbn/config-schema';
+import { RouteDefinitionParams } from '.';
+import { SAMLLoginStep } from '../authentication';
+
+export function defineAuthenticationRoutes(params: RouteDefinitionParams) {
+  if (params.config.authc.providers.includes('saml')) {
+    defineSAMLRoutes(params);
+  }
+}
+
+/**
+ * Defines routes required for SAML authentication.
+ */
+function defineSAMLRoutes({
+  basePath,
+  router,
+  logger,
+  authc,
+  getLegacyAPI,
+}: RouteDefinitionParams) {
+  function createCustomResourceResponse(body: string, contentType: string) {
+    return {
+      body,
+      headers: {
+        'content-type': contentType,
+        'cache-control': 'private, no-cache, no-store',
+        'content-security-policy': getLegacyAPI().cspRules,
+      },
+      statusCode: 200,
+    };
+  }
+
+  router.get(
+    {
+      path: '/api/security/authc/saml/capture-url-fragment',
+      validate: false,
+      options: { authRequired: false },
+    },
+    (context, request, response) => {
+      const currentBasePath = basePath.get(request);
+      // We're also preventing `favicon.ico` request since it can cause new SAML handshake.
+      return response.custom(
+        createCustomResourceResponse(
+          `
+          <!DOCTYPE html>
+          <title>Kibana SAML Login</title>
+          <link rel="icon" href="data:,">
```

Comment:
__note:__ alternatively we can make`favicon.ico` or/and anything that ends with `.ext` as special case in `canRedirectRequest`....

This becomes a problem only if we store `redirectPath` in the cookie when we render this custom HTML `favicon.ico` request may cause us to create a new cookie :/

Reply:
> For static resources like the favicon.ico, we don't require auth. 

We don't, but we also run auth code before we figure out that the request is targeting non-existent resource (as it's done by browser automatically if we don't define favicon, `somehost/favicon.ico`) and for security this request looks like any other non-resource request. But now I started wondering if it's expected that we call `auth` hook for the 404 route....

> It looks like we're using the equivalent of http://localhost:5601/ui/favicons/favicon.ico normally, should we be doing this here instead?

I considered this option, but this `ui/favicons` path is maintained in another place and if someday they change the path this code will silently be broken (unless we try to simulate SAML IdP and write a functional test for that). Ideally this should be solved with https://github.com/elastic/kibana/issues/41964 and until then we can I either use current workaround or switch to `ui/favicons/favicon.ico` and hope that it won't change before https://github.com/elastic/kibana/issues/41964. Both options sound good to me, what would you prefer?

Reply:
> I don't mind us merging with the current work-around also.

We still have time, but good to know! :)